ksql-engine/src/main/java/io/confluent/ksql/metrics/ConsumerCollector.java

```diff
@@ -0,0 +1,202 @@
+ * </p>
+ * Potential issues:
+ * - With metric leaks where a client interceptor is closed(), De-registration of sensors and metrics where other clients are still connected (in the same process).
+ * - Around dynamic resource allocation (ie. stream threads are added/removed)
+ * </p>
+ * Benefits:
+ * - Smaller memory footprint due to agg on intercept
+ */
+public class ConsumerCollector implements MetricCollector {
+
+  private final Map<String, Counter> topicPartitionCounters = new HashMap<>();
+  private Metrics metrics;
+  private String id;
+
+  public void configure(Map<String, ?> map) {
+    String id = (String) map.get(ConsumerConfig.GROUP_ID_CONFIG);
+    configure(MetricCollectors.addCollector(id, this), id);
+  }
+
+  ConsumerCollector configure(final Metrics metrics, final String id) {
+    this.id = id;
+    this.metrics = metrics;
+    return this;
+  }
+
+  @Override
+  public String getId() {
+    return id;
+  }
+
+  public ConsumerRecords onConsume(ConsumerRecords records) {
+    collect(records);
+    return records;
+  }
+
+  @SuppressWarnings("unchecked")
+  private void collect(ConsumerRecords consumerRecords) {
+    Stream<ConsumerRecord> stream = StreamSupport.stream(consumerRecords.spliterator(), false);
+    stream.forEach((record) -> topicPartitionCounters.computeIfAbsent(getKey(record.topic().toLowerCase()), k ->
+            new Counter<>(record.topic().toLowerCase(), buildSensors(k))
+    ).increment(record));
+  }
+
+  private String getKey(String topic) {
+    return topic;
+  }
+
+  private Map<String, Counter.SensorMetric<ConsumerRecord>> buildSensors(String key) {
+
+    HashMap<String, Counter.SensorMetric<ConsumerRecord>> results = new HashMap<>();
+
+    // Note: synchronized due to metrics registry not handling concurrent add/check-exists activity in a reliable way
+    synchronized (this.metrics) {
+      addRateSensor(key, results);
+      addBandwidthSensor(key, results);
+      addTotalSensor(key, results);
+    }
+    return results;
+  }
+
+  private void addRateSensor(String key, HashMap<String, Counter.SensorMetric<ConsumerRecord>> results) {
+    String name = "cons-" + key + "-rate-per-sec";
+
+    //noinspection unchecked
+    MetricName metricName = new MetricName("consume rate-per-sec", name, "consumer-rate-per-sec",  Collections.EMPTY_MAP);
+    Sensor existingSensor = metrics.getSensor(name);
+    Sensor sensor = metrics.sensor(name);
+
+    // re-use the existing measurable stats to share between consumers
+    if (existingSensor == null) {
+      sensor.add(metricName, new Rate(TimeUnit.SECONDS));
+    }
+
+    KafkaMetric rate = metrics.metrics().get(metricName);
+    results.put(metricName.name(), new Counter.SensorMetric<ConsumerRecord>(sensor, rate) {
+      void record(ConsumerRecord record) {
+        sensor.record(1);
+        super.record(record);
+      }
+    });
+  }
+
+  private void addBandwidthSensor(String key, HashMap<String, Counter.SensorMetric<ConsumerRecord>> results) {
```

Comment:
This method and the next one look pretty similar - anyway we can share the duplicate code?

Reply:
ok - done that - dropped bandwidth measurement on the consumer because the anony-class impl uses the consumerrecord.serializedbytes instead of passing in 1